cli/internal/run/global_hash.go

```diff
@@ -47,28 +47,40 @@ type oldGlobalHashable struct {
 }
 
 // calculateGlobalHashFromHashable returns a hash string from the globalHashable
-func calculateGlobalHashFromHashable(named GlobalHashable) (string, error) {
-	// When we aren't in infer mode, we can hash the whole object
-	if named.envMode != util.Infer {
-		return fs.HashObject(named)
-	}
+func calculateGlobalHashFromHashable(full GlobalHashable) (string, error) {
+	switch full.envMode {
+	case util.Infer:
+		if full.envVarPassthroughs != nil {
+			// In infer mode, if there is any passThru config (even if it is an empty array)
+			// we'll hash the whole object, so we can detect changes to that config
+			// Further, resolve the envMode to the concrete value.
+			full.envMode = util.Strict
+			return fs.HashObject(full)
+		}
 
-	// In infer mode, if there is any passThru config (even if it is an empty array)
-	// we'll hash the whole object, so we can detect changes to that config
-	if named.envVarPassthroughs != nil {
-		return fs.HashObject(named)
+		// If we're in infer mode, and there is no global pass through config,
+		// we use the old struct layout. this will be true for everyone not using the strict env
+		// feature, and we don't want to break their cache.
+		return fs.HashObject(oldGlobalHashable{
+			globalFileHashMap:    full.globalFileHashMap,
+			rootExternalDepsHash: full.rootExternalDepsHash,
+			envVars:              full.envVars.All.ToHashable(),
+			globalCacheKey:       full.globalCacheKey,
+			pipeline:             full.pipeline,
+		})
+	case util.Loose:
+		// Remove the passthroughs from hash consideration if we're explicitly loose.
+		full.envVarPassthroughs = nil
+		return fs.HashObject(full)
+	case util.Strict:
+		// Collapse `nil` and `[]` in strict mode.
+		if full.envVarPassthroughs == nil {
+			full.envVarPassthroughs = make([]string, 0)
+		}
+		return fs.HashObject(full)
+	default:
+		panic("unimplemented environment mode")
```

Comment:
- What's the benefit of using `panic` over bubbling up an error?
- Is it even possible to hit this case, if we're limiting the CLI flag to a set of known values?

Reply:
Bubbling up an error to the user implies that they have an ability to do something about this issue. They don't.

`panic` is for early notice of code authoring errors.

In general this pattern is the primitive version of exhaustiveness checking. "We blow up if you don't account for all your inputs." Rust has compiler-enforced exhaustiveness checking while Go only has a third-party linter.

This is presently impossible to hit but is designed to blow up if you add a new mode and don't implement it in the switch.

Reply:
In other words: our tests should hit this and then blow up if you or I failed to write this code.